toolchain/check/merge.cpp

```diff
@@ -233,7 +233,14 @@ static auto CheckRedeclParam(Context& context, bool is_implicit_param,
 
   auto prev_param_type_id = SemIR::GetTypeInSpecific(
       context.sem_ir(), prev_specific_id, prev_param_pattern.type_id());
-  if (!context.types().AreEqualAcrossDeclarations(new_param_pattern.type_id(),
+  if ((!allow_self_type_mismatch ||
+       (SemIR::Function::GetNameFromPatternId(context.sem_ir(),
+                                              new_param_pattern_id) !=
+            SemIR::NameId::SelfValue ||
+        SemIR::Function::GetNameFromPatternId(context.sem_ir(),
```

Comment:
Sorry, this is bothering me as I think about how this works. Lines 280-284 are fundamentally about querying the parameter name. Is there a way that the result of the name fetch could be reused for this logic?

Reply:
Note, not sure if it's easier to merge with my above comment that the shape of `self` probably should match... i.e., maybe we can just move the type comparison to be after the `addr` unwrap and name comparison?

Reply:
Yeah, rolled that into the other change ( c82572a10cf118b5a6d4385965af5b619055e8c8 ).